docs/transient-crr-source.md

```diff
@@ -0,0 +1,184 @@
+  a transient location, and returning errors to PUT requests if the
+  limit is reached.
+
+## Design
+
+### Design overview
+
+CRR process reads from the source defined by the location constraint,
+writes to one or more destinations asynchronously and updates statuses
+to *COMPLETED* on the source when done.
+
+Transient source relies on this core behavior, but refers to a
+temporary primary location for a new object, until the CRR completes
+the processing of this object.
+
+The following then happens when an object is put to a location which
+is configured as being a transient source:
+
+* the object is first put to the location which is configured as a
+  transient source, then the put status is returned as usual,
+
+* CRR processors transfer the object asynchronously to all CRR targets
+  set through replication API. It sends an extra header in the GET
+  request "x-amz-location-constraint" to ensure reads are done from
+  the transient source location, not from the *preferred read
+  location*,
+
+* finally, once all CRR targets have been processed, the transient
+  data is deleted asynchronously by the garbage collector backbeat
+  service. Reads from users still happen on the *preferred read
+  location*.
+
+If for some reason the CRR cannot be completed to some backends, the
+data will be held on the transient source, and the data will still be
+read from the *preferred read location* if CRR was successful on
+it. If CRR is either PENDING or FAILED to the preferred read target,
+it's still possible to provide the "x-amz-location-constraint" header
+on GET requests, specifying the location from which to read the data.
+
+Once CRR has eventually completed successfully to all backends (either
+through automatic or manually-triggered retries), the transient source
+data is garbage-collected asynchronously, and the object will not be
+readable anymore from the transient source. We do not update object
+metadata to reflect the fact that the data on transient source has
+effectively been deleted, the user must be aware that only the CRR
+cloud targets remain valid sources.
+
+### Detailed design
+
+Here are the changes per component:
+
+#### CloudServer
+
+* support for "isTransient" property on locations: enforces that
+  replication configuration of buckets that belong to transient
+  locations specify a *preferred read location*
+
+* new backbeat route `/_/backbeat/batchDelete`, that is used by the
+  garbage collector process to delete data locations. It is used
+  through POST requests containing a JSON body that is an array of
+  locations to delete.
+
+* modified support of *preferred read locations*: now it's set through
+  the replication configuration of a bucket, by appending the
+  `:preferred_read` suffix to the given location in the
+  `<StorageClass>` attribute. This has support from the Orbit user
+  interface through the locations setup.
+
+* added storage limit support through the use of Redis keys storing
```

Comment:
Just curious, are these Redis keys stored permanently in sentinel and removed as needed?

Reply:
They are never removed right now.

Dora raised an interesting point that if we re-create a location later with the same name, we would re-use the same keys. Maybe this is a strong incentive to cleanup when we delete locations, or alternatively making sure the keys do not conflict when re-creating locations (she should be opening a ticket for this shortly).